x-pack/plugins/apm/public/components/shared/FilterBar/DatePicker.tsx

```diff
@@ -0,0 +1,220 @@
+  public state = {
+    from: '',
+    to: '',
+    isPaused: true,
+    refreshInterval: 0
+  };
+  public lastRefresh = 0;
+
+  public componentDidMount() {
+    // read from query string to initialize, or set defaults
+    const g = this.getG(this.props.location.search);
+    this.updateStateFromUrl(g);
+  }
+
+  public componentDidUpdate(prevProps: Props) {
+    if (prevProps.location.search !== this.props.location.search) {
+      const currentG = this.getG(this.props.location.search);
+      const previousG = this.getG(prevProps.location.search);
+      if (currentG !== previousG) {
+        this.updateStateFromUrl(currentG);
+      }
+    }
+  }
+
+  public getG(search: string) {
+    const { _g } = toQuery(search);
+    return risonSafeDecode(_g) as QueryG | undefined;
+  }
+
+  public refresh = () => {
+    const { from, to, isPaused, refreshInterval } = this.state;
+    const now = new Date().getTime();
+    const parsed = {
+      from: datemath.parse(from),
+      to: datemath.parse(to)
+    };
+
+    // UNCOMMENT THIS FOR DEBUGGING WHILE IN DEVELOPMENT, REMOVE BEFORE MERGING
+    // console.log('refreshing', {
+    //   last: this.lastRefresh,
+    //   isPaused,
+    //   refreshInterval,
+    //   since: now - this.lastRefresh || now,
+    //   parsed
+    // });
+
+    if (!parsed.from || !parsed.to || isPaused || !refreshInterval) {
+      // TODO: Can we do anything better here if parsed values are undefined?
+      return;
+    }
+    if (!this.lastRefresh) {
+      this.lastRefresh = now;
+    }
+    if (now - this.lastRefresh >= refreshInterval) {
+      // UNCOMMENT THIS FOR DEBUGGING IN DEV, REMOVE BEFORE MERGING
+      // console.log(
+      //   '%cACTUAL REFRESH UPDATE',
+      //   'color: red; font-size: 30px; padding: 20px 0'
+      // );
+      const min = parsed.from.toISOString();
+      const max = parsed.to.toISOString();
+      this.props.dispatchUpdateTimePicker({ min, max });
+      this.lastRefresh = now;
+    }
+    requestAnimationFrame(this.refresh);
+  };
+
+  public updateStateFromUrl(g: QueryG = {}, previousG: QueryG = {}) {
+    const { from, to } = g.time || APM_DEFAULT_TIME_RANGE;
+    const { pause, value } = g.refreshInterval || APM_DEFAULT_REFRESH;
+
+    const parsed = {
+      from: datemath.parse(from),
+      to: datemath.parse(to)
+    };
+
+    if (!parsed.from || !parsed.to) {
+      // TODO: Can we do anything better here?
+      return;
+    }
+
+    const min = parsed.from.toISOString();
+    const max = parsed.to.toISOString();
+    this.props.dispatchUpdateTimePicker({ min, max });
+    this.setState({ from, to, isPaused: pause, refreshInterval: value }, () => {
+      this.refresh();
+    });
+  }
+
+  public updateUrlG(updatedG: QueryG) {
+    const nextSearch = getQueryWithRisonParams(this.props.location, '', {
+      _g: updatedG
+    });
+
+    this.props.history.replace({
+      ...this.props.location,
+      search: fromQuery(nextSearch)
+    });
+  }
```

Comment:
I think we should very seriously consider getting rid of rison inside APM UI. It will simplify url parsing a lot. We still need rison when handling urls to Discover but we can treat that as legacy.

Reply:
Are we sure there aren't any incoming links that we care about breaking if we don't read from the rison on load, at least? Does Kibana really want every app handling date/time in their own way with different query params in every app?

Reply:
> Are we sure there aren't any incoming links that we care about breaking if we don't read from the rison on load, at least?

I can't think of other plugins than Infra that links to us. We should mention this to them.

> Does Kibana really want every app handling date/time in their own way with different query params in every app?

In an ideal world, no. But I think we should use this opportunity to simplify something that has bothered us for a while. And then we can hopefully affect the rest of the platform over time.

Reply:
@makwarth do you have any opinion on this? Options:

1. Leave the URL format as it works today, which means reading from/writing to the URL in this "rison" format which involves some unnecessary complexity around encoding and decoding the time range and refresh values. Any inbound links into APM that specify date ranges would continue to work the same way the do today.

1. Begin using our own query parameters outside of the existing `_g` param, without the "rison" format. Simplifies our code reading from and writing to the URL. Downside is we could break inbound links, and it increases the complexity of the work being done in this ticket a bit. We also need to pick a new set of query parameters which would be nice to coordinate with other Kibana apps, but there may not be time to do that coordination at this point in the release cycle. 

Thoughts? I think we should definitely get rid of the rison complexity, but I wasn't sure if that was something we wanted to do right now in this ticket.

Reply:
I imagine we'd want to go with option 1 for now, for the risks pointed out in option 2, but also because of two reasons;
 
1. We only want to use the UI component and not bring with it additional refactoring of the time handling if possible. 

2. I assume there'll be changes incoming from the platform team when/if they're addressing the date and time handling. Also assuming we're not the only team dependent on the rison format?

Reply:
> 1. We only want to use the UI component and not bring with it additional refactoring of the time handling if possible.

My point is that Rison has a lot of complexity, which is visible in this PR. I'm convinced that this PR could be smaller, if we dropped the Rison dep.

> 2\. I assume there'll be changes incoming from the platform team when/if they're addressing the date and time handling. Also assuming we're not the only team dependent on the rison format?

Just to clear up a potential misunderstanding: The Timepicker UI Component or the platform doesn't prescribe any specific date format. We can choose to use Rison encoded values for time range, as well as for all the other values, or not. Eg. ML went full-on Rison for all values. We have only used Rison for date ranges, and normal encoding for other values.

Reply:
@jasonrhodes I was only able to find this inbound link to APM, and it doesn't contain time range:

https://github.com/elastic/kibana/blob/00fbc2572efe519dd2169b3e21845b1397ecdf21/x-pack/plugins/infra/public/components/waffle/node_context_menu.tsx#L54-L63

Reply:
@sqren My response was merely to says that I'm in favour of leaving things a little more complex until we can perhaps refactor things to depend more or less on Rison for another release. I'm not sure I understand why there's no suggested way to handle date formatting in Kibana, but I'm also sure someone once gave me the answer to this and I've forgotten it (when we first implemented APM UI).

Reply:
> until we can perhaps refactor things to depend more or less on Rison for another release

That would be in 8.0 to avoid breaking the urls. I don't think I can handle Rison in APM UI that long 😅

Reply:
Rison has been removed -- let me know if anyone objects. :) Resolving this for now.